File: ios/DailyMemory/Domain/Repository/PersonRepository.swift
import Foundation
import Combine

/// Repository protocol for Person operations.
/// Defines the contract for data access, implemented in data layer.
protocol PersonRepository {
    // CRUD Operations
    func save(_ person: Person) async throws
    func update(_ person: Person) async throws
    func delete(id: String) async throws

    // Query - Single
    func getById(_ id: String) async throws -> Person?

    // Query - All (with different sorting)
    func getAll(sortedBy: PersonSortOrder) async throws -> [Person]

    // Query - By Relationship
    func getByRelationship(_ relationship: Relationship) async throws -> [Person]

    // Query - Search
    func search(query: String) async throws -> [Person]

    // Query - Recently Met
    func getRecentlyMet(limit: Int) async throws -> [Person]

    // Statistics
    func getCount() async throws -> Int
    func incrementMeetingCount(id: String, date: Date) async throws
}

/// Sort order for persons
enum PersonSortOrder {
    case alphabetical
    case recentMeeting
    case mostFrequent
}

File: android/app/src/main/java/com/dailymemory/domain/repository/PersonRepository.kt
package com.dailymemory.domain.repository

import com.dailymemory.domain.model.Person
import com.dailymemory.domain.model.Relationship
import com.dailymemory.domain.model.SyncStatus
import kotlinx.coroutines.flow.Flow
import java.time.LocalDateTime

/**
 * Repository interface for Person operations.
 * Defines the contract for data access, implemented in data layer.
 */
interface PersonRepository {

    // CRUD Operations
    suspend fun insert(person: Person)
    suspend fun update(person: Person)
    suspend fun delete(person: Person)
    suspend fun deleteById(id: String)

    // Query - Single
    suspend fun getById(id: String): Person?
    fun observeById(id: String): Flow<Person?>

    // Query - All (with different sorting)
    fun observeAll(): Flow<List<Person>>
    fun observeAllByLastMeeting(): Flow<List<Person>>
    fun observeAllByMeetingCount(): Flow<List<Person>>

    // Query - By Relationship
    fun observeByRelationship(relationship: Relationship): Flow<List<Person>>

    // Query - Search
    fun searchByName(query: String): Flow<List<Person>>

    // Query - Recently Met
    fun observeRecentlyMet(limit: Int): Flow<List<Person>>

    // Query - Not Contacted
    fun observeNotContactedSince(sinceDate: LocalDateTime): Flow<List<Person>>

    // Query - Sync
    suspend fun getUnsyncedPersons(): List<Person>
    suspend fun updateSyncStatus(id: String, status: SyncStatus)

    // Statistics
    suspend fun getCount(): Int
    suspend fun incrementMeetingCount(id: String, date: LocalDateTime)
}
